textmap/utilities.py

```diff
@@ -148,12 +144,205 @@ def fit_transform(self, X, y=None, **fit_params):
+            j = indices[idx]
+            bg_val = 0.0
+            for k in range(background_i.shape[1]):
+                bg_val += background_i[i, k] * background_j[k, j]
+            row_background[idx] = bg_val
+
+        row_background = row_background / row_background.sum()
+
+        mix_param = 0.5
+        current_dist = mix_param * row_data + (1.0 - mix_param) * row_background
+
+        last_estimated_dist = mix_param * current_dist + (1.0 - mix_param)
+
+        change_vec = last_estimated_dist
+        change_magnitude = 1.0 + precision
+
+        while (
+                change_magnitude > precision and mix_param > 1e-2 and mix_param < 1.0 - 1e-2
+        ):
+
+            posterior_dist = (current_dist * mix_param) / (
+                    current_dist * mix_param + row_background * (1.0 - mix_param)
+            )
+
+            current_dist = posterior_dist * row_data
+            mix_param = (current_dist.sum() + prior[0]) / mp
+            current_dist = current_dist / current_dist.sum()
+
+            estimated_dist = mix_param * current_dist + (1.0 - mix_param)
+            change_vec = np.abs(estimated_dist - last_estimated_dist)
+            change_vec /= estimated_dist
+            change_magnitude = np.sum(change_vec)
+
+            last_estimated_dist = estimated_dist
+
+            # zero out any small values
+            norm = 0.0
+            for n in range(current_dist.shape[0]):
+                if current_dist[n] < low_thresh:
+                    current_dist[n] = 0.0
+                else:
+                    norm += current_dist[n]
+            current_dist /= norm
+
+        result[indptr[i]: indptr[i + 1]] = current_dist
+        mix_weights[i] = mix_param
+
+    return result, mix_weights
+
+
+def multinomial_em_sparse(
+        matrix, background_i, background_j, precision=1e-4, low_thresh=1e-5, bg_prior=5.0
+):
+    result = matrix.tocsr().copy().astype(np.float32)
+    new_data, mix_weights = numba_multinomial_em_sparse(
+        result.indptr,
+        result.indices,
+        result.data,
+        background_i,
+        background_j,
+        precision,
+        low_thresh,
+        bg_prior,
+    )
+    result.data = new_data
+
+    return result, mix_weights
+
+
 class RemoveEffectsTransformer(BaseEstimator, TransformerMixin):
+    """
+
+       Parameters
+       ----------
+       n_components: int
+           The target dimension for the low rank model (will be exact under pLSA or approximate under EnsTop).
+
+       model_type: string
+           The model used for the low-rank approximation.  To options are
+           * 'pLSA'
+           * 'EnsTop'
+
+        optional EM params:
+        * precision = 1e-4,
+        * low_thresh = 1e-5,
+        * bg_prior = 5.0,
+
+       """
+
+    def __init__(self, n_components=1, model_type="pLSA", **em_params):
+
+        self.n_components = n_components
+        self.model_type = model_type
+        self.em_params = em_params
```

Comment:
Since there aren't that many params it is better to explicitly have them has named attributes in the model.
```suggestion
        self.em_threshold = em_threshold
        self.em_background_prior = em_background_prior
        self.em_precision = em_precision
```

Reply:
Agreed

Reply:
Cool. I agree